lib/codegen-meta/src/base/types.rs

```diff
@@ -0,0 +1,262 @@
+//! This module predefines all the Cretonne scalar types.
+
+use std::fmt;
+
+// Numbering scheme for value types:
+//
+// 0: Void
+// 0x01-0x6f: Special types
+// 0x70-0x7f: Lane types
+// 0x80-0xff: Vector types
+//
+// Vector types are encoded with the lane type in the low 4 bits and log2(lanes)
+// in the high 4 bits, giving a range of 2-256 lanes.
+static LANE_BASE: u8 = 0x70;
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum Bool {
+    /// 1-bit bool.
+    B1 = 1,
+    /// 8-bit bool.
+    B8 = 8,
+    /// 16-bit bool.
+    B16 = 16,
+    /// 32-bit bool.
+    B32 = 32,
+    /// 64-bit bool.
+    B64 = 64,
+}
+
+impl Bool {
+    /// Get the number of a boolean variant.
+    pub fn number(self) -> u8 {
+        let offset = match self {
+            Bool::B1 => 0,
+            Bool::B8 => 1,
+            Bool::B16 => 2,
+            Bool::B32 => 3,
+            Bool::B64 => 4,
+        };
+
+        LANE_BASE + offset
+    }
+}
+
+pub struct BoolIterator {
+    index: usize,
+}
+
+impl BoolIterator {
+    pub fn new() -> Self {
+        Self { index: 0 }
+    }
+}
+
+impl Iterator for BoolIterator {
+    type Item = Bool;
+    fn next(&mut self) -> Option<Self::Item> {
+        let res = match self.index {
+            0 => Some(Bool::B1),
+            1 => Some(Bool::B8),
+            2 => Some(Bool::B16),
+            3 => Some(Bool::B32),
+            4 => Some(Bool::B64),
+            _ => None,
+        };
+        self.index += 1;
+        res
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum Int {
+    /// 8-bit int.
+    I8 = 8,
+    /// 16-bit int.
+    I16 = 16,
+    /// 32-bit int.
+    I32 = 32,
+    /// 64-bit int.
+    I64 = 64,
+}
+
+impl Int {
+    /// Get the number of an integer variant.
+    pub fn number(self) -> u8 {
+        let offset = 5 + match self {
+            Int::I8 => 0,
+            Int::I16 => 1,
+            Int::I32 => 2,
+            Int::I64 => 3,
+        };
+
+        LANE_BASE + offset
+    }
+}
+
+pub struct IntIterator {
+    index: usize,
+}
+
+impl IntIterator {
+    pub fn new() -> Self {
+        Self { index: 0 }
+    }
+}
+
+impl Iterator for IntIterator {
+    type Item = Int;
+    fn next(&mut self) -> Option<Self::Item> {
+        let res = match self.index {
+            0 => Some(Int::I8),
+            1 => Some(Int::I16),
+            2 => Some(Int::I32),
+            3 => Some(Int::I64),
+            _ => None,
+        };
+        self.index += 1;
```

Comment:
What about release mode overflow? (Call `.next()` `usize::MAX` times and you get `I8` again) You could replace the `_ => None,` with `_ => return None,` to prevent this.